ts/identifier_pool.go

```diff
@@ -39,59 +38,83 @@ type simpleIdentifierPool struct {
 	pool pool.ObjectPool
 }
 
-func (p *simpleIdentifierPool) GetBinaryID(ctx context.Context, v []byte) ID {
+func (p *simpleIdentifierPool) GetBinaryID(ctx context.Context, v checked.Bytes) ID {
 	id := p.pool.Get().(*id)
 	ctx.RegisterFinalizer(id)
 
-	id.data = v
+	v.IncRef()
+	id.pool, id.data = p, v
 
 	return id
 }
 
 func (p *simpleIdentifierPool) GetStringID(ctx context.Context, v string) ID {
-	return p.GetBinaryID(ctx, []byte(v))
+	data := checked.NewBytes([]byte(v), nil)
+	return p.GetBinaryID(ctx, data)
 }
 
 func (p *simpleIdentifierPool) Put(v ID) {
 	v.Reset()
 	p.pool.Put(v)
 }
 
-func (p *simpleIdentifierPool) Clone(other ID) ID {
+func (p *simpleIdentifierPool) Clone(existing ID) ID {
 	id := p.pool.Get().(*id)
-	id.data = other.Data()
+
+	data := existing.Data()
+	data.IncRef()
+
+	newData := checked.NewBytes(nil, nil)
+	newData.IncRef()
+	newData.AppendAll(data.Get())
+
+	data.DecRef()
+
+	id.pool, id.data = p, newData
 
 	return id
 }
 
 // NewNativeIdentifierPool constructs a new NativeIdentifierPool.
 func NewNativeIdentifierPool(
-	heap pool.BytesPool, options pool.ObjectPoolOptions) IdentifierPool {
-
-	if options == nil {
-		options = pool.NewObjectPoolOptions()
+	heap pool.CheckedBytesPool,
+	opts pool.ObjectPoolOptions,
+) IdentifierPool {
+	if opts == nil {
+		opts = pool.NewObjectPoolOptions()
 	}
 
-	return &nativeIdentifierPool{
-		pool: pool.NewNativePool(pool.NativePoolOptions{
-			Type: reflect.TypeOf(id{}),
-			Size: uint(options.Size()),
-		}), heap: configureHeap(heap)}
+	iopts := opts.InstrumentOptions()
+
+	p := &nativeIdentifierPool{
+		pool: pool.NewObjectPool(opts.SetInstrumentOptions(
+			iopts.SetMetricsScope(iopts.MetricsScope().SubScope("id-pool")))),
+		heap: configureHeap(heap),
+	}
+	p.pool.Init(func() interface{} {
+		return &id{pool: p}
+	})
+	return p
 }
 
 type nativeIdentifierPool struct {
-	pool pool.NativePool
-	heap pool.BytesPool
+	pool pool.ObjectPool
```

Comment:
A comment or two here explaining why we switched to normal pooling would make it easier to comprehend the rationale behind it.